modules/standard/Set.chpl

```diff
@@ -347,26 +374,28 @@ module Set {
       :yields: A reference to one of the elements contained in this set.
     */
     iter these() {
-      for x in _dom.these() do
-        yield x;
+      for idx in 0..#_htb.tableSize do
+        if _htb.isSlotFull(idx) then yield _getSlotFromIdx(idx).key;
     }
 
     pragma "no doc"
     iter these(param tag) where tag == iterKind.standalone {
-      for x in _dom.these(tag) do
-        yield x;
+      // TODO: Does this need to be a forall?
+      var chunk = 0..#_htb.tableSize;
+      for idx in chunk.these(tag) do
+        if _htb.isSlotFull(idx) then yield _getSlotFromIdx(idx).key;
     }
 
     pragma "no doc"
     iter these(param tag) where tag == iterKind.leader {
-      for followThis in _dom.these(tag) do
-        yield followThis;
+      var chunk = 0..#_htb.tableSize;
+      for followThis in chunk.these(tag) do yield followThis;
     }
 
     pragma "no doc"
     iter these(param tag, followThis) where tag == iterKind.follower {
-      for x in _dom.these(tag, followThis) do
-        yield x;
+      for idx in followThis do
+        if _htb.isSlotFull(idx) then yield _getSlotFromIdx(idx).key;
```

Comment:
The leader/follower don't seem right to me.  They'll work with each other, but won't play well with other leaders/followers, IIUC.

Given say a set `mySet` with `_htb.tableSize` 53 and `_htb.tableNumFilledSlots` 25 (say) and one task, the leader will yield `0..52`, and the follower will follow `0..52` and yield 25 things.

But if paired with a `zip(0..52, mySet)`, the range leader will yield `0..52` (ok), but the range follower will yield `0..52` whereas the set follower will yield only 25 elements.  Or `zip(0..24, mySet)`, the range leader will yield `0..24`, and the range follower will yield `0..24`, but the set follower will yield however many elements are set in the first 25 slots of the hash table.

If you turn the `zip()` around, things work out poorly the other way around too.

I'd think the leader has to yield `0..#htb.numFilledSlots`, and the follower following `(a..#c)` unfortunately has to count `a` filled slots in and then yield the next `c`.

(I'd also think this iteration is something the hash table should implement.)

Though maybe I'm missing something.  I can't figure out how the current `DefaultAssociativeDom` (using the same hashtable module) can be zipped with anything other than not only a `DefaultAssociativeDom`, but a domain with a hash table that has precisely the same slots filled.  (And indeed, `forall (i,j) in zip(D,r)` (and vice versa) where `D` is a `domain(int)` fails to compile (and fails to compile differently).  So maybe `set`s don't need to be zippable with anything other than `set`s?

Reply:
That's a really good point, how much work each task is doing depends on the distribution of the keys...

So yeah I think it's correct to say that in the worst case you could have a distribution so poor that one task ends up getting "all" the filled slots and thus, does all the work. And that gets even worse when you consider two different sets might have completely different distributions of keys.

Can you think of an option to avoid this that doesn't allocate extra storage when iterating?

Reply:
Note it's not just a load imbalance issue -- if two followers on the same task on the same iteration (given the same followThis) yield different numbers of iterands, I believe that will result in a runtime `halt()`.  (The same as a `forall` over a zip of two 1-D arrays of different sizes.)

This was my thought as to what to do:

> I'd think the leader has to yield `0..#htb.numFilledSlots`, and the follower following `(a..#c)` unfortunately has to count `a` filled slots in and then yield the next `c`.

The idea is to treat the `followThis` yielded by the leader and accepted by the followers as a dense range, where the follower will yield an element for every index in the range.  (So across all the tasks, the combined sizes of all the `followThese` should be `numFilledSlots`.)

The leader would need to split the range up if there's more than one task, which I didn't include in the quote.  For the follower, I should have written "skip the first `a` filled slots ...".

I think that would make sets compatible with ranges and 1D domains and arrays, and with the DynamicIters compatible with them, etc.

I also think this should be implemented by the hash table, and set could delegate its iteration to that implementation.

Reply:
> But even then, I think this leader/follower pair will be able to reliably zip two sets only if those sets are the same set.

The DefaultAssociative code handles this by looking up the element from the main set in the followed set.

I'm not really convinced we need to solve this problem now in this PR, but I think there might be an interesting solution yielding only full slots in the leader. I just don't think we have it quite yet, because the current implementation is sacrificing scalability.

Reply:
> The DefaultAssociative code handles this by looking up the element from the main set in the followed set.

I think that means a DefaultAssociative domain can follow only another DefaultAssociative domain (or array?) that has the same indices in the same slots, right?

For a set, that would mean (I think) that a set could be zippered reliably only with a set with the same members -- e.g., mathematically the same set.  Either as a follower as a leader.  Is that useful?

All I can think of to be able to follow a filled-slot-number is for the hash table to maintain a List (or suchlike) of filled slot numbers, so the follower can simply index into it with the values from followThis.  Which seems like a fair amount of overhead especially if the set is never actually iterated over.

Reply:
Unless we can require the caller to manage that filled-slot-number List or Array for us,

```chapel
var fsna = set..buildFilledSlotNumberArray(); // passed on to htbl
forall (i,j) in zip(1..#set.size, set.these(fsna)) {
  ...
}
```

Then `set.these(tag == follower, fsna = [], followThis)` can just yield `_htbl.table[fsna[t]]` for each t in its followThis.  Or if fsn is empty (user didn't bother), use the current inefficient algorithm.
(I don't see a way for the `set` follower to do this without its caller's help.)


If that makes sense, maybe "leader/follower 2.0" could allow for `proc someIter(tag == followerInit)` that could return an `object?` to be passed to an `iter someIter(tag == follower)`?

Reply:
We can return to this discussion in a separate effort :).